src/liballoc/allocator.rs

```diff
@@ -0,0 +1,698 @@
+    /// calls to those methods.
+    ///
+    /// However, for clients that do not wish to track the capacity
+    /// returned by `alloc_excess` locally, this method is likely to
+    /// produce useful results.
+    unsafe fn usable_size(&self, layout: &Layout) -> (Capacity, Capacity) {
+        (layout.size(), layout.size())
+    }
+
+    // == METHODS FOR MEMORY REUSE ==
+    // realloc. alloc_excess, realloc_excess
+
+    /// Returns a pointer suitable for holding data described by
+    /// `new_layout`, meeting its size and alignment guarantees. To
+    /// accomplish this, this may extend or shrink the allocation
+    /// referenced by `ptr` to fit `new_layout`.
+    ///
+    /// * `ptr` must have previously been provided via this allocator.
+    ///
+    /// * `layout` must *fit* the `ptr` (see above). (The `new_layout`
+    ///   argument need not fit it.)
+    ///
+    /// * `new_layout` must have size greater than zero.
+    ///
+    /// Behavior undefined if any of the above three constraints are unmet.
+    ///
+    /// In addition, `new_layout` should not impose a different alignment
+    /// constraint than `layout`. (In other words, `new_layout.align()`
+    /// should equal `layout.align()`.)
+    /// However, behavior is well-defined (though underspecified) when
+    /// this constraint is violated; further discussion below.
+    ///
+    /// If this returns `Ok`, then ownership of the memory block
+    /// referenced by `ptr` has been transferred to this
+    /// allocator. The memory may or may not have been freed, and
+    /// should be considered unusable (unless of course it was
+    /// transferred back to the caller again via the return value of
+    /// this method).
+    ///
+    /// Returns `Err` only if `new_layout` does not meet the allocator's
+    /// size and alignment constraints of the allocator or the
+    /// alignment of `layout`, or if reallocation otherwise fails. (Note
+    /// that did not say "if and only if" -- in particular, an
+    /// implementation of this method *can* return `Ok` if
+    /// `new_layout.align() != old_layout.align()`; or it can return `Err`
+    /// in that scenario, depending on whether this allocator
+    /// can dynamically adjust the alignment constraint for the block.)
+    ///
+    /// If this method returns `Err`, then ownership of the memory
+    /// block has not been transferred to this allocator, and the
+    /// contents of the memory block are unaltered.
+    unsafe fn realloc(&mut self,
+                      ptr: Address,
+                      layout: Layout,
+                      new_layout: Layout) -> Result<Address, AllocErr> {
+        let new_size = new_layout.size();
+        let old_size = layout.size();
+
+        if new_size > old_size {
+            if let Ok(()) = self.grow_in_place(ptr, layout.clone(), new_layout.clone()) {
+                return Ok(ptr);
+            }
+        } else if new_size < old_size {
```

Comment:
Oh yeah, good point. I know I was also non-plussed by the handling of same size inputs, but cannot remember why I didn't just resolve it in the manner you suggest. (Apart from just being unused to the new `grow`/`shrink` methods and possibly forgetting if they required the new size to be strictly greater/less than the old one...)